### ISSUE TITLE
Support defining enabled and disabled lints in a configuration file

### ISSUE THREAD

**Issue by Rantanen** [rust-lang/cargo#5034]
_This issue started its life in clippy (https://github.com/rust-lang-nursery/rust-clippy/issues/1313). Once it was realized this should apply to built-in lints as well, it migrated to rust (https://github.com/rust-lang/rust/issues/45832). There @kennytm suggested that it should be Cargo's responsibility to pass the `-D`/`-W`/`-A` flags to rustc._

### Motivation

Currently project-wide lint configuration needs to be included in the crate sources. This is okay when the project consists of a single crate, but gets more cumbersome when the project is a workspace with dozen crates.

Being able to define the lints in an external file that will be used when building the crates would have several benefits:

- Sharing one configuration file for all the crates in a workspace.
- A canonical place to check for lint configuration when contributing to a new project.
- An easy way to examine the version history for lint changes.

### Proposal

- A configuration `lints.toml` (or `[lints]` in `Cargo.toml`?) file that specifies the lints.
- The lints can be specified on the workspace level and for individual packages. Anything on the package level will override the workspace setup on per lint basis.
- The configuration file is cfg/feature-aware and can specify different lints depending on features. Specifying clippy-lints will result in clippy complaining about unknown lints if clippy isn't used.

Also... unlike what @oli-obk suggested in the rust issue, personally I'd prefer the following format:

```toml
[lints]
allow_unused = "allow"
non_snake_case = "allow"
```

While this is more verbose than `allow = [ "allow_unused", "non_snake_case" ]`, I think the diff benefits and the ability for the user to group logically similar lints together even if some of them are `deny` and some `allow` outweighs the increased verbosity.

Also if Cargo/rustc ever support lint configurations, this would be more future proof:

```toml
cyclomatic_complexity = { state = "allow", threshold = 30 }
```

### Example configuration

```toml
[lints]
allow_unused = "deny"
non_snake_case = "allow"

[lints.'cfg(feature = "clippy")']
cyclomatic_compexity = "deny"
```

The feature syntax is lifted from Cargo's dependency syntax - at least that was the intention. Personally I'm not a fan of it, but I guess that's a limitation of toml. Don't really have a better proposal for it, given I don't really know what's available int he toml syntax. :<

**Comment by Mark-Simulacrum** [comment]
I am strongly in favor. We should also probably specify that unknown lints should be a warning, not an error, for the purposes of backwards compatibility.

**Comment by Rantanen** [comment]
Forgot to mention: I'm also interested in contributing code for the implementation, if there is consensus on what should happen. Also do Cargo changes, such as this, go through the RFC process?

**Comment by Mark-Simulacrum** [comment]
This seems potentially worthy of an RFC, but I could see it going either way. There doesn't seem to be much design space. @alexcrichton can probably say more.

**Comment by alexcrichton** [comment]
This seems plausible to me! I think it'd be fine to add this to `Cargo.toml` on an unstable basis (behind a nightly feature) and we can decide to stabilize at a later date.

**Comment by lukaslueg** [comment]
Do we have a chance for `rustc` to report where the linter settings are coming from? Given the increasing amount of knobs and switches, it would be very nice to have `rustc` report that `allow_unused` was denied and that `deny` was specified by `the workspace's root Cargo.toml` or `the commandline`.

As `rustc` knows nothing about these, we could introduce a new switch to `rustc` that specifies the source of the following linter settings (e.g. `--LS "/home/myproject/foobar/Cargo.toml" -D ... --LS "/home/myproject/Cargo.toml" -D ...`).

Is this a thing?

**Comment by oli-obk** [comment]
I'm not sure if these should be in the Cargo.toml, as we many want to be able to specify them globally for a group of projects.

**Comment by oli-obk** [comment]
> Is this a thing?

rustc already reports whether something was

* defined by command line
* defined by default settings
* defined by code

so we just add another variant for config file there.

**Comment by phansch** [comment]
I'm really interested in this, too.

> Also if Cargo/rustc ever support lint configurations, this would be more future proof:
>     `cyclomatic_complexity = { state = "allow", threshold = 30 }`

It could be `cyclomatic_complexity = { state = "allow" }` from the start, then we would have the option to add further options in a second step.

**Comment by phansch** [comment]
This Rollup PR from rust demonstrates how rust itself could benefit from a shared lint config, too: https://github.com/rust-lang/rust/pull/52268 (note all the 'deny bare trait object' PRs; and there's a few more)

**Comment by detrumi** [comment]
I'd like to start working on this, unless @phansch or @Rantanen started on this already.

Seems like the `lints.toml` name is the one I've seen used most, so let's go with that.
As for the syntax, `allow_unused = "allow"` seems reasonable and more ergonomic than requiring the `cyclomatic_complexity = { state = "allow" }` syntax.

> I'm not sure if these should be in the Cargo.toml, as we many want to be able to specify them globally for a group of projects.

A `Cargo.toml` file can already be used for a workspace, unless you meant a group of workspaces?

**Comment by Mark-Simulacrum** [comment]
I'd prefer that we centralize this into Cargo.toml rather than adding an additional file, but that can likely be decided in further detail at a later point in time.

**Comment by oli-obk** [comment]
The argument against `Cargo.toml` is that you can have a single `lints.toml` for multiple projects. Or even hierarchical `lints.toml`s

**Comment by phansch** [comment]
@detrumi I didn't start any work on this, and won't have that much time either. From my side, feel free to start :+1:

**Comment by Mark-Simulacrum** [comment]
@oli-obk Right, which is what workspaces are for IMO. I'd prefer to avoid "one more" file in every Rust directory.

**Comment by oli-obk** [comment]
That would mean that companies need one enormous workspace for all the crates in their company if they want to have a consistent lint setup. This is the very use case that started the discussion about lint configuration files. A company-wide uniform linting experience. Maybe specializeable for crates, but they can already do that via `#![allow(foo)]` in their `lib.rs`. If we leave the door open for importing lint configurations in a dependency-style manner, doing it in `Cargo.toml` seems fine though.

**Comment by Luthaf** [comment]
Why not use `.cargo/config` here ? This seems like the obvious candidate for a file shared across repositories/workspaces/crates. Plus, one can commit a `.cargo/config` file per repository in OSS scenarios.

**Comment by detrumi** [comment]
Alright, placing it in `Cargo.toml` seems like a better place to start then.

> That would mean that companies need one enormous workspace for all the crates in their company if they want to have a consistent lint setup.

Hierarchical `Cargo.toml` files might be a solution to the enormous workspaces, but that's something that can be added later.

**Comment by Rantanen** [comment]
> I'd like to start working on this, unless @phansch or @Rantanen started on this already.

Feel free to! I have been waiting for some kind of a decision on how to proceed, so haven't done anything for this myself.

**Comment by sanmai-NL** [comment]
@oli-obk, @Mark-Simulacrum: My company has one enormous workspace indeed. And we are highly in need of DRY linting/warnings/etc. configuration with local exceptions (preferably down to the statement-level) So if industrial use cases/productivity are the motivation, perhaps we can survey a bit as for the requirements?

**Comment by oli-obk** [comment]
> down to the statement-level

that part is no problem, you can configure lints in source code just fine.

And if you have one enormous workspace, then the `Cargo.toml` solution will work just fine for you I think.

**Comment by sanmai-NL** [comment]
Suppose we settle on configuring this in `Cargo.toml`, then in a later stage we can design a more general feature, namely supporting `Cargo.toml` override files. Such override files could e.g. patch sections such as the lints and warnings configuration. We can then circumvent the discussion about the need for a separate configuration file. The name of the section is more important, since compiler warnings and clippy lints may be merged in some form later.

**Comment by marrub--** [comment]
Is this dead currently? For projects with a hierarchical structure and lots of lints this is a fairly vital feature. Hoping to see this gain activity soon!

**Comment by ehuss** [comment]
@marrub-- A more detailed proposal was posted at https://internals.rust-lang.org/t/proposal-cargo-lint-configuration/9135.

**Comment by joshtriplett** [comment]
I very much don't want to see this go into a separate configuration file. I've worked with C and similar for decades, where warning options live in the build system rather than in source code, and that leads to quite a bit of skew where source code doesn't have the information needed to build.

I *definitely* think this information should not be in `.cargo/config` or anywhere that needs separate setup, and it shouldn't be in a workspace's Cargo.toml if that information doesn't get published in a .crate file. I could live with `Cargo.toml`, though I'd prefer sticking with code (and having a dependency crate that encompasses a pile of options for an include-like mechanism).

**Comment by dbeckwith** [comment]
The main thing I'm hoping to avoid is the repetition of lints in every crate in a workspace. I don't much care about it being in a config file vs. source code as long as there's some way to inherit/include the list of lints from some source.

**Comment by oli-obk** [comment]
Maybe we could come up with a scheme where you can import the lint configuration from another crate? So making it a language feature where you can define a crate that does nothing but set lint flags, and then you have another crate depend on it and import these flags into its own scope.

**Comment by sanmai-NL** [comment]
@joshtriplett: I do not understand your argument. The proposal is not to remove static analysis configuration from Rust source files, but to also support setting this configuration in an external file. 

You mention being able to build from source alone, rather than source complemented with packaging metadata. Presently, Rust source code in the open source ecosystem of crates.io often cannot be built from source alone. The configuration in `Cargo.toml` is rather crucial. I define building from source alone as just using `rustc` and then input and appropriate output files, across many command lines.

**Comment by dbeckwith** [comment]
@sanmai-NL I think part of his argument was that some of those configuration files don't end up being published with the crate, especially if you're in a workspace. So having the lints live in a top-level `Cargo.toml` or something means that it wouldn't be included in individual crate sources, and information required to build the source properly would be missing.
I don't know how often people try to build crates from the source published on crates.io though; I would always go to the original repository to make sure I'm getting the full source code.

**Comment by sanmai-NL** [comment]
@dbeckwith: But even if the configuration isn't being included in the source distribution via crates.io,  builds won't fail because of it. Clippy does not break builds when run as standalone tool. In the past there was a compiler plugin, but I do not see documentation about it in the current [README.md](https://github.com/rust-lang/rust-clippy/blob/c7d444508655eab132f0d24d81f9cdc08b848056/README.md). Now suppose there is some possibility that Clippy is used within `rustc`'s compilation pipeline and that the configuration to it is the only way to unbreak this pipeline. Only then, for under those conditions, Clippy could be designed to not accept out-of-source configuration. That would solve the problem.

The use case for this feature request is of course within a context of Clippy being run separately, during a build pipeline (Continuous X).

Also, [`clippy.toml` already exists](https://github.com/rust-lang/rust-clippy/tree/c7d444508655eab132f0d24d81f9cdc08b848056#configuration) as a configuration mechanism. So @joshtriplett's concern is not specific to the proposal here.

**Comment by dbeckwith** [comment]
@sanmai-NL What about `#deny` rules? If there are any of those, then building with vs. without clippy would provide different results in regards to whether the code will build successfully or not, which is important for repeatability.
I agree though, maybe this point is not relevant to this feature request, and your point about `clippy.toml` is a good one.

**Comment by sanmai-NL** [comment]
Suppose the scenario @joshtriplett tries to get at happens: the external config hasn't been distributed together with the source code.

* If there are Clippy deny rules in the code, these will be enforced as before.
* If there are Clippy deny rules in the original external configuration, these will not impede the build, since the external configuration is absent now.

So what @dbeckwith describes isn't a valid failure scenario.

The only valid failure scenario I can think of is the following. There are `allow` rules that except _default_ `deny` rules. The `allow` rules are (only) present in external configuration, which itself is absent. 

There are, however, no default `deny` rules, according to https://rust-lang.github.io/rust-clippy/master/index.html. Only within the `clippy::correctness` lint group, which isn't enabled by default. So this scenario is currently impossible.

Now let's concede the following conditions: the maintainership of some software solution in Rust has enabled the `clippy::correctness` lint group, and then made exceptions to that in external configuration. You want to build this solution to a release artifact using the Rust compiler only. Let's also concede that a Clippy compiler plugin is configured in the source code to be run unconditionally. Let's furthermore concede that a release build does not [cap lints](https://doc.rust-lang.org/rustc/lints/levels.html#capping-lints). Only then, perhaps, we have a valid failure scenario.

I find this wildly far-fetched and at so many steps the maintainers and consumers would be going against best practices, that I find it hard to see @joshtriplett reservation as warranted.

Not to mention, by making one or more simple design decisions we can defuse the scenario.  As I wrote in my previous comment: Clippy could be designed to not accept out-of-source configuration under some conditions, for example when there is Clippy configuration within the source code already.

**Comment by joshtriplett** [comment]
On March 21, 2019 6:02:09 AM PDT, Sander Maijers <notifications@github.com> wrote:
>@dbeckwith: But even if the configuration isn't being included in the
>source distribution via crates.io,  builds won't fail because of it.
>Clippy does not break builds when run as standalone tool. In the past
>there was a compiler plugin, but I do not see documentation about it in
>the current
>[README.md](https://github.com/rust-lang/rust-clippy/blob/c7d444508655eab132f0d24d81f9cdc08b848056/README.md).
>Now suppose there is some possibility that Clippy is used within
>`rustc`'s compilation pipeline and that the configuration to it is the
>only way to unbreak this pipeline. Only then, for under those
>conditions, Clippy could be designed to not accept out-of-source
>configuration. That would solve the problem.
>
>The use case for this feature request is of course within a context of
>Clippy being run separately, during a build pipeline (Continuous X).
>
>Also, [`clippy.toml` already
>exists](https://github.com/rust-lang/rust-clippy/tree/c7d444508655eab132f0d24d81f9cdc08b848056#configuration)
>as a configuration mechanism. So @joshtriplett's concern is not
>specific to the proposal here. 

Note that this isn't just about clippy, it's also about compiler lints.

**Comment by joshtriplett** [comment]
To clarify something: I can live with lint configuration in the `Cargo.toml` shipped with a crate (though based on experience with many C packages I feel that will still cause problems). However, lint configuration in a file *not* shipped in the `.crate` file will cause problems for distributions. That includes shipping lint configuration in some other `.toml` file that Cargo doesn't ship, or shipping lint configuration in a configuration file of a workspace, or shipping lint configuration in `.cargo/config`.

**Comment by peterjoel** [comment]
I think it is important that there is support for a separate `lints.toml` file, even if lint configuration is added to `Cargo.toml`. This will make it much easier to share common linting standards across multiple projects.

* At a minimum, the lint configuration from a workspace needs to be inherited by child crates.
* Ideally, it should be possible to create a single lint configuration outside the project and just create a symlink to it within the project itself. This would be tricky to do if configuration was only supported in `Cargo.toml`.

**Comment by joshtriplett** [comment]
> Ideally, it should be possible to create a single lint configuration outside the project and just create a symlink to it within the project itself.

That is exactly the kind of situation I'd like to avoid.

At a minimum, Cargo would need to understand the mechanism in question well enough to undo it when uploading a `.crate` file, so that the crate is fully self-contained.

**Comment by peterjoel** [comment]
That makes sense. However, there is still a need to share common lint settings across projects. The exact mechanism isn't importan,  but the obvious workaround to use `include!` in the root of every crate, isn't very good either.

**Comment by Nemo157** [comment]
> To clarify something: I can live with lint configuration in the `Cargo.toml` shipped with a crate (though based on experience with many C packages I feel that will still cause problems). However, lint configuration in a file _not_ shipped in the `.crate` file will cause problems for distributions. That includes shipping lint configuration in some other `.toml` file that Cargo doesn't ship, or shipping lint configuration in a configuration file of a workspace, or shipping lint configuration in `.cargo/config`.

Given `--cap-lints` is used when compiling dependencies, is there any need for the lint config to be part of the published package? It feels like it is really only relevant during development, once published users only care whether the crate compiles or not.

I guess a binary that `allow`s a `deny`-by-default lint wouldn't compile without it, but distributions should probably be building with `--cap-lints` specified on the top-level themselves to avoid failing to build with new future-compat lints anyway.

**Comment by carols10cents** [comment]
[This RFC for deduplicating Cargo workspace information](https://github.com/rust-lang/rfcs/pull/2906) looks relevant to configuring lints in Cargo.toml.

**Comment by rkuhn** [comment]
I share @joshtriplett’s concern about self-contained .crate files, but I don’t see the logical implication that lint config must be defined within the packaged source: can the effective lint settings not be included in the .crate file during packaging? One mechanism to do so would be to have a `[lint]` section in `Cargo.toml`, which is transformed by the packaging process to include settings that came from other sources at the location where the crate was built. Making this process reproducible is important, so I’d favour configuration within the source repo, for example in the workspace’s `Cargo.toml`. This is a lot less hassle than requiring the ceremony of remembering to add a bunch of text to `lib.rs` whenever starting a new crate.

**Comment by RazgrizHsu** [comment]
are there any update for lint config?

**Comment by jmaargh** [comment]
I'd like to help get this solved if I can.

Having read through the above, it seems like there's a bit of a tension between (a) the view of "lint as developer aid" and (b) that of "lint as compiler flag".

For (a), I don't think the issue of lint config making crates not self-contained is an issue at all, since consumers of the crate aren't it's developers and (as said above) "users only care whether the crate compiles or not". Whereas (b) does seem to imply that lint config needs to be contained within the shipped crate.

In my experience (a) is what's more commonly understood by the idea of "lints". This more naturally leads to a much more liberal configuration scheme where organisations, workspaces, users, etc. can all define and customise the lint scheme to suit themselves during development (perhaps with `lint.toml` files searched for up the filesystem tree as mentioned above). CI can then enforce a "canonical" set of denied lints and everyone gets on with their day.

The only thing I've noticed in the above discussion which seems to force the (b) view is the idea that there may be "deny by default" lints, thus implying that not shipping lint config could break the build of a crate. The idea of a "deny by default" lint initially feels somewhat oxymoronic to me, but I'm happy to be informed if I'm wrong on that.

For what it's worth, my initial feeling is to forbid the concept of "deny by default" lints and let people have as many `lint.toml` files in as many places as they'd like.

tl;dr: a sensible config scheme seems to depend on what we think "lint" means exactly (is it the same as a compiler warning? a compilation option? something else?) Please let me know what I can do to move things forward.

**Comment by workingjubilee** [comment]
> The only thing I've noticed in the above discussion which seems to force the (b) view is the idea that there may be "deny by default" lints, thus implying that not shipping lint config could break the build of a crate. The idea of a "deny by default" lint initially feels somewhat oxymoronic to me, but I'm happy to be informed if I'm wrong on that.

Deny by default lints are common for rustc's future incompatibility lints: They move from a gradual progression of allow -> warn -> deny -> hard error. In addition, arithmetic overflow is an example of a deny-by-default lint.
https://doc.rust-lang.org/rustc/lints/listing/deny-by-default.html

**Comment by jmaargh** [comment]
I'll admit that I don't fully understand all of the issues here, but given that `--lint-caps=allow` is used when compiling dependencies, doesn't that entirely solve the deny-by-default issue?

**Comment by Manishearth** [comment]
Yeah I do not see what the issue with deny by default is. Dependencies are built with lint levels capped, this is no different from lint levels being specified in the source.

Ideally this can be specified at the workspace level too.

**Comment by workingjubilee** [comment]
I was not aware of that detail! So for dependencies they're probably not at all a concern. I was mostly just expressing they're useful as an intermediate step between "this is totally invalid and doesn't make sense" and "this almost certainly shouldn't actually compile, but I can compile it if you _really_ want me to".

Which, on that note: I think whatever is arrived at should probably offer a good alternative to the current [`#![deny(warnings)]` antipattern](https://rust-unofficial.github.io/patterns/anti_patterns/deny-warnings.html).

**Comment by Manishearth** [comment]
I'm actually not sure if that antipattern still applies given that cap-lints exists. The patterns book is old and partially maintained, and cap-lints was not always a thing.

Again, if a crate is a dependency there is no way to cause it to fail to compile by tweaking lint levels.

It's not a huge deal if a crate fails to compile under deny(warnings), if you're building the crate itself you probably are developing on it (not using it as a dep). Except perhaps for binary crates.

**Comment by sanmai-NL** [comment]
@workingjubilee: [I don’t think it’s an antipattern at all](https://github.com/rust-unofficial/patterns/issues/46#issuecomment-345215058).

**Comment by est31** [comment]
One big problem with `#![deny(warnings)]` is that you have to add it to lib.rs, main.rs, all your example .rs files, otherwise you might not notice warnings in CI, if that's your intent of `#![deny(warnings)]`. At that point it's way easier to add `-D warnings` to `RUSTFLAGS`. Ideally you'd have a central file that cargo reads the lints from, which this issue is about, because then it doesn't just apply to CI but local builds as well.

**Comment by jmaargh** [comment]
Am I missing something, or is the discussion of `#![deny(warnings)]` entirely tangential to this issue?

AFAICT, so long as `--lint-caps=allow` is used when compiling dependency crates _or_ shipped binary crates (e.g. with `cargo install`), then a lint config file (e.g. `lint.toml`) described as above can just be added with no breaking changes. It clearly seems to be a desirable feature.

**Comment by Manishearth** [comment]
Yes, it is, as I've pointed out earlier.

**Comment by workingjubilee** [comment]
Erm, @est31's remark is all that I meant, as effectively I was merely noting it as another desirable, as someone who has in fact had to tediously go through every crate in a large workspace and clean up `#![deny(warnings)]` (and `#![deny(warnings, ...)]`, `#![deny(..., warnings, ...)]`, and `#![deny(..., warnings)]`) from each one in order to make it manageable to fix up a crate, and would have preferred to have one line to diff, so I would be very happy about there being an obvious feature for that? ^^;

**Comment by Manishearth** [comment]
Oh, I see what you mean. Sorry, I was looking at the "reasons why this is an antipattern" on the patterns repo you linked to and noting they were outdated. Yes, it would definitely be nice to centralize that effort.

**Comment by eholk** [comment]
I've been playing around with different ways to do this. You can get part of the way there by using `rustflags` in `.cargo/config.toml`. The problem is, that doesn't understand the `clippy::` lints, since those are only installed in `rustc` when run through `clippy-driver`. So, I wonder if we could have `clippy-driver` set a `clippy` cfg setting, and then we could add the following to `config.toml`:

```
[target.'cfg(clippy)']
rustflags = [
  "-Dclippy::foo",
  "-Aclippy::bar",
]
```

It's maybe not as pretty as adding `lints.toml`, but it seems like it'd be a small change that'd get something that works. What do people think? Are there cases I'm missing with this approach?

**Comment by Manishearth** [comment]
> The problem is, that doesn't understand the `clippy::` lints, since those are only installed in `rustc` when run through `clippy-driver`.

My understanding of the design of `clippy::` was that rustc should _ignore_ them when passed through without clippy, so if it's not doing that that's a bug in rustc.

**Comment by eholk** [comment]
> My understanding of the design of `clippy::` was that rustc should _ignore_ them when passed through without clippy, so if it's not doing that that's a bug in rustc.

Hmm, interesting. When I try to specify clippy lints with command line options I get the following:

```
> rustc -A clippy::imprecise_flops .\src\main.rs
error[E0602]: unknown lint: `clippy::imprecise_flops`
  |
  = note: requested on the command line with `-A clippy::imprecise_flops`
```

I assumed this was intentional, but it is weird that this behavior is different than with the `#[allow(clippy::foo)]` attributes.

**Comment by Manishearth** [comment]
I'd say it's unintentional

**Comment by eholk** [comment]
Oh yeah, it seems like it is. It looks like we don't parse out the lint tool name for lints given on the command line argument.

I have a fix for it at https://github.com/eholk/rust/commit/e0c3fdb0ff34f3323f542b5e4ae71aee3350c380. I'm going to clean it up a little more in the morning and then I'll submit a PR for it.

**Comment by gilescope** [comment]
Is there a 3rd party cargo- clippy wrapping crate that fixes this _now_ while we work out a perfect solution to put forward as unstable?

**Comment by gilescope** [comment]
I see someone's solved this: https://crates.io/crates/cargo-lints . Good stuff. Ttfn.

**Comment by gilescope** [comment]
Re-reading everything, it looks like https://github.com/rust-lang/cargo/pull/9684 will be very impactful (as has been previously mentioned) - maybe we should wait till that lands?

Meanwhile, now that the PR ignoring clippy rust flags has landed, the workaround works with a slight change: the clippy feature is named cargo-clippy ( https://github.com/rust-lang/rust-clippy/pull/1057 ).

The following `.cargo/config.toml` seems to work fine:

```
[target.'cfg(feature = "cargo-clippy")']
rustflags = [
  "-Dclippy::foo",
  "-Aclippy::bar",
]
```
Can we use this as a reasonable workaround for now?
(cargo-lints works fine but a built-in method is preferable)

**Comment by camsteffen** [comment]
Trying to piece together what the end result might look like...

1. Lint levels may be specified in a crate's metadata, either in `[lints]` or `lints.toml`.
2. A crate should be able to optionally publish it's lint config to be used by other crates
    ```toml
    [lib]
    publish_lint_config = true
    ```
    or should it be a crate type? (or maybe a broader crate type like `"metadata"`?)
     ```toml
     [lib]
     crate_type = "lint-config"
     ```
3. Lint configs from other crates should be import-able
   ```toml
   [dev-dependencies]
   my_lint_config = "*"
   
   [lints]
   includes = ["my_lint_config"]
   [lints.levels]
   "clippy::pedantic" = "warn"
    ```

Can we pursue #​1 independently? And can we defer the `lints.toml` idea for now? It seems like `[lints]` would be a good MVP, and is wanted with or without the option of `lints.toml`.

**Comment by camsteffen** [comment]
Thinking more about `lints.toml`. It seems like the main concern with it is that it creates too much decoupling of lint config from a crate's metadata. That is, lint config should be considered a component of crate metadata because it influences how/whether it compiles. This is a conflict of interest because people want lint config to be decoupled so that it can be shared between multiple crates. The solution is allow cargo to facilite this sharing through dependencies/workspaces rather than using a symlink or some filesystem-based solution. This is different from `rustfmt.toml` because formatting has no influence on how a crate is compiled.

We could theoretically still support `lints.toml` as an "extension" to `Cargo.toml`, perhaps with something like `[lints.external] = true`, but I think it's better not to do this for simplicity.

**Comment by camsteffen** [comment]
> Would a `[lints]` section be usable at a workspace-level?

I think that's where https://github.com/rust-lang/cargo/issues/8415 comes in.

**Comment by wagenet** [comment]
Setting with `[target.'cfg(feature = "cargo-clippy")']` as @gilescope recommended works but it has a very large downside of causing `cargo clippy` and `cargo build` to clobber each other's build caches since the rustflags have changed.

**Comment by MarijnS95** [comment]
@wagenet We've so far worked around that by using `[target.'cfg(all())'] rustflags = ...`: https://github.com/EmbarkStudios/rust-ecosystem/pull/68#issuecomment-945863477

**Comment by ivan-aksamentov** [comment]
Would be nice to take inspiration from older ecosystems, like ESLint. JS/TS community has one of the best dev experiences with regards to static analysis, and I believe they "solved" linting.
 - https://eslint.org/
 - Config example: https://github.com/facebook/react/blob/4175f05934d3954827c00b7633c1ce8e4e87d227/.eslintrc.js#L35-L123

In the `.eslintrc.js` file they have a `rules` object, where the keys are the names of the lints. They allow any string/object/array as value and simply pass this value into the corresponding rule implementation. This way, the config stays unopinionated and does not restrict rule implementations in any way. This also allows to skip the discussion of what is "the best" config layout - let the rule implementors to decide!

Additionally, there are some nice features that reduce friction of creating new lints, and of sharing them with the community:
 - ESLint allows `plugins`, the packages which users can install to add new rules. Example: [eslint-plugin-react](https://github.com/jsx-eslint/eslint-plugin-react).
 - The `extends` section allows reusing existing configs, for example configs used by large, well-known companies, or configs tailored for specific environments (e.g. in Rust, embedded community can have their own lints). These configs can be packaged and installed as well. Example: [eslint-config-airbnb](https://github.com/airbnb/javascript/tree/master/packages/eslint-config-airbnb)

For me, a Rust newcomer with extensive experience in other languages, cargo check/clippy is a huge pain to configure, even though the lints themselves are fantastic.

When I dive into a new ecosystem, I tend to learn its best practices from lints, because they are a dense, summarized form of the essential knowledge, carefully crafted by the experienced community members. I feel like these gurus helping me out and teaching me, while I am writing code. However, sometimes clippy is configured to be to shy, even if I am eager to learn about more strict rules. But then I need to paste a huge blob of attributes in every library,  binary, test and example to enable this knowledge. This is ridiculously bad. 

I believe Rust should be making more emphasis on dev experience related to lints to help newcomers.

**Comment by lnicola** [comment]
> But then I need to paste a huge blob of attributes in every library, binary, test and example to enable this knowledge. This is ridiculously bad.

I don't see this workaround suggested here, but you can use a `cargo` alias (in your home directory, maybe) to set a preferred invocation. Here's an example: https://github.com/rust-lang/rust-analyzer/blob/5008e5682159965c6f3b174aed3523c0b5e2fc08/.cargo/config#L6=.

**Comment by ivan-aksamentov** [comment]
@lnicola Thanks! This is not a bad workaround and would work for CLI usage. Will not work for text editors though, and is of course nowhere near the great dev experience that I believe Rust is  providing in other aspects.

**Comment by Kinrany** [comment]
> Thinking more about `lints.toml`. It seems like the main concern with it is that it creates too much decoupling of lint config from a crate's metadata. That is, lint config should be considered a component of crate metadata because it influences how/whether it compiles. This is a conflict of interest because people want lint config to be decoupled so that it can be shared between multiple crates.

We could flatten the lint configs into a single file when publishing the crate. Does that help?

**Comment by ehaynes99** [comment]
@ivan-aksamentov Note that eslint is in the middle of completely reworking their configuration.

https://github.com/eslint/eslint/issues/13481

**Comment by ivan-aksamentov** [comment]
@ehaynes99 Interesting, thanks! Not sure what it's prompted by. I am quite happy with the existing ESLint config. But will delay my judgement until something comes up.

**Comment by thomaseizinger** [comment]
Over in https://github.com/rust-lang/rust/issues/54726, [someone mentioned](https://github.com/rust-lang/rust/issues/54726) the idea of using custom inner attributes to expand to a set of linting rules.

Could this possibly be a complete replacement for this feature? Sharing a lint config across a workspace could then be implemented as a proc-macro that reads a configuration file, be it a company-wide file hosted somewhere or locally within the workspace.

The only downside would be that the proc-macro needs to be included in every crate but that seems acceptable to me for the benefit of implementing this entire idea through an existing feature. Proc-macros can already be published to crates-io so we additionally get the whole "publish and reuse lints from other people" for free.

**Comment by ivan-aksamentov** [comment]
@thomaseizinger 
> Sharing a lint config across a workspace could then be implemented as a proc-macro that reads a configuration file

I'd consider this a clever hack. And in fact that's how I learn that macro cannot contain attributes - I tried to write one. It would be a copy-pasting machine that hides the defect. Also it's easy to forget to add the macro to a new workspace package, or accidentally remove it, and then it will not not be properly linted.

Why it was decided that the whole linter config should be in attirubutes? It does not belong there. That's where things went wrong and that's what needs to be fixed IMO.

Linter is a tool. Tools need configs. Clippy is one of the Rust linters. I love it. But why is that Clippy so special it requires my source code to be polluted with it all over the place?

Also you cannot setup the precise options for rules in the attributes, at least currently. i.e. I want cyclomatic complexity 200, instead of default. There is currently no way to do it with attributes, so there is clippy.toml file anyways, but it's almost unused and detached from the real config.

**Comment by Manishearth** [comment]
We're not. There's general consensus that we should _do_ this here, there's just not enough work being done on design and implementation.

I'd ask people to limit this kind of decision so that this issue can be used to actually talk about designing said feature. The pointless discussions and "me too" comments do not help and tend to make it more likely that anyone on the teams working on this stuff will mute the thread.

**Comment by flip1995** [comment]
I started writing up a RFC with the MVP @joshtriplett suggests. I never got to finishing it due to limited time. I plan to work on this in the near future though. However, I'm aware about the feature freeze of `cargo`. But I think we should get the ball rolling with making a community accepted design choice (should it be a config file, included in `Cargo.toml`, a mix of both, ...?) through an RFC.

**Comment by Muscraft** [comment]
if `Cargo.toml` is chosen it might be a good idea to extend [workspace inheritance](https://github.com/rust-lang/cargo/blob/master/src/doc/src/reference/unstable.md#workspace-inheritance). It would make sharing `lints` easier. 

If that is the case I think `allow = [ "allow_unused", "non_snake_case" ]` would be better and make toml parsing easier. Not having to make a field for each lint would be much better overall. Ideally, the `Cargo.toml` would look like below

```toml
# [PROJECT_DIR]/Cargo.toml
[workspace.lints]
allow = [ "allow_unused", "non_snake_case" ]
```

```toml
# [PROGJCT_DIR]/bar/Cargo.toml
[lints]
workspace = true
deny = [ "allow_unused" ]
```

**Comment by CobaltCause** [comment]
I don't know if this has been brought up before, but what about something a little more general like [Python's `pyproject.toml` `[tool]` section from PEP 518](https://peps.python.org/pep-0518/#tool-table)? After having worked with Python for a little over a year, I found this rather pleasant to use since it means all tools (that support it) use the same format, in the same file, which helps with uniformity, ease-of-use, and reduces file clutter. [Cargo already has `[package.metadata]` and `[workspace.metadata]`](https://doc.rust-lang.org/cargo/reference/manifest.html#the-metadata-table) too which sound pretty similar.

**Comment by ParkMyCar** [comment]
Hey @flip1995, you mentioned you started on an RFC? I'd be more than happy to help, or pickup the work if you've become too busy?

**Comment by flip1995** [comment]
Yes, I started one some time ago, but can't find it anymore. :facepalm: There is this attempt at an RFC: https://hackmd.io/@flip1995/HJteAAPgO But it is really sparse and not at a point where I would continue with it, but rather start fresh.

**Comment by emilk** [comment]
For anyone looking for a workaround for this issue, I have been using `cargo cranky` for the past six months, and I'm really liking it: https://github.com/ericseppanen/cargo-cranky

**Comment by epage** [comment]
Since #12115 is merged, I'm closing this as resolved.
